idx/memory/memory.go

```diff
@@ -168,52 +185,60 @@ func (t *TagIndex) delTagId(name, value string, id schema.MKey) {
 	}
 }
 
-// org id -> nameWithTags -> Set of references to schema.MetricDefinition
-// nameWithTags is the name plus all tags in the <name>;<tag>=<value>... format.
-type defByTagSet map[uint32]map[string]map[*schema.MetricDefinition]struct{}
+// org id -> NameWithTagsHash() -> Map keyed by *idx.MetricDefinition.
+type defByTagSet map[uint32]map[idx.Md5Hash]map[*idx.MetricDefinition]struct{}
 
-func (defs defByTagSet) add(def *schema.MetricDefinition) {
-	var orgDefs map[string]map[*schema.MetricDefinition]struct{}
+func (defs defByTagSet) add(def *idx.MetricDefinition) {
+	var orgDefs map[idx.Md5Hash]map[*idx.MetricDefinition]struct{}
 	var ok bool
 	if orgDefs, ok = defs[def.OrgId]; !ok {
-		orgDefs = make(map[string]map[*schema.MetricDefinition]struct{})
+		orgDefs = make(map[idx.Md5Hash]map[*idx.MetricDefinition]struct{})
 		defs[def.OrgId] = orgDefs
 	}
 
-	fullName := def.NameWithTags()
-	if _, ok = orgDefs[fullName]; !ok {
-		orgDefs[fullName] = make(map[*schema.MetricDefinition]struct{}, 1)
+	hashedName := def.NameWithTagsHash()
+	if _, ok = orgDefs[hashedName]; !ok {
+		orgDefs[hashedName] = make(map[*idx.MetricDefinition]struct{}, 1)
+	}
+
+	if _, ok = orgDefs[hashedName][def]; !ok {
+		orgDefs[hashedName][def] = struct{}{}
 	}
-	orgDefs[fullName][def] = struct{}{}
 }
 
-func (defs defByTagSet) del(def *schema.MetricDefinition) {
-	var orgDefs map[string]map[*schema.MetricDefinition]struct{}
+func (defs defByTagSet) del(def *idx.MetricDefinition) {
+	var orgDefs map[idx.Md5Hash]map[*idx.MetricDefinition]struct{}
 	var ok bool
 	if orgDefs, ok = defs[def.OrgId]; !ok {
 		return
 	}
 
-	fullName := def.NameWithTags()
-	delete(orgDefs[fullName], def)
+	hashedName := def.NameWithTagsHash()
+	delete(orgDefs[hashedName], def)
 
-	if len(orgDefs[fullName]) == 0 {
-		delete(orgDefs, fullName)
+	if len(orgDefs[hashedName]) == 0 {
+		delete(orgDefs, hashedName)
 	}
 
 	if len(orgDefs) == 0 {
 		delete(defs, def.OrgId)
 	}
 }
 
-func (defs defByTagSet) defs(id uint32, fullName string) map[*schema.MetricDefinition]struct{} {
-	var orgDefs map[string]map[*schema.MetricDefinition]struct{}
+func (defs defByTagSet) defs(id uint32, fullName string) map[*idx.MetricDefinition]struct{} {
+	var orgDefs map[idx.Md5Hash]map[*idx.MetricDefinition]struct{}
 	var ok bool
 	if orgDefs, ok = defs[id]; !ok {
 		return nil
 	}
 
-	return orgDefs[fullName]
+	md5Sum := md5.Sum(bytes.NewBufferString(fullName).Bytes())
+	hashedName := idx.Md5Hash{
+		Upper: binary.LittleEndian.Uint64(md5Sum[:8]),
+		Lower: binary.LittleEndian.Uint64(md5Sum[8:]),
+	}
+
+	return orgDefs[hashedName]
```

Comment:
this logic relies on the assumption that there will never be a hash collision. it would be easy to deal with the collision by taking `defs := orgDefs[hashedName]`, looping over `defs` and comparing the `fullName` to each of them before returning. In most cases when there's no collision this might waste a few cpu cycles, but it will prevent that wrong results get returned due to a collision, so i think that's probably worth it.